### PR TITLE
Create voltage init modifications

### PR DESCRIPTION
**Pull request by FranckLecuyer**

### CODE REVIEW

File: src/main/java/org/gridsuite/modification/server/dto/ModificationInfos.java [hunk 1]
```diff
@@ -35,33 +35,34 @@
...
+    @JsonSubTypes.Type(value = LineAttachToVoltageLevelInfos.class),
+    @JsonSubTypes.Type(value = LinesAttachToSplitLinesInfos.class),
+    @JsonSubTypes.Type(value = BranchStatusModificationInfos.class),
+    @JsonSubTypes.Type(value = EquipmentAttributeModificationInfos.class),
+    @JsonSubTypes.Type(value = GeneratorScalingInfos.class),
+    @JsonSubTypes.Type(value = LoadScalingInfos.class),
+    @JsonSubTypes.Type(value = DeleteVoltageLevelOnLineInfos.class),
+    @JsonSubTypes.Type(value = DeleteAttachingLineInfos.class),
+    @JsonSubTypes.Type(value = GenerationDispatchInfos.class),
+    @JsonSubTypes.Type(value = MassiveEquipmentsModificationsInfos.class)
```
Comment (jonenst) [inline src/main/java/org/gridsuite/modification/server/dto/ModificationInfos.java:65]: why this change ?

File: src/main/java/org/gridsuite/modification/server/entities/equipment/modification/MassiveEquipmentsModificationsEntity.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        super.update(massiveEquipmentsModificationsInfos);
+        assignAttributes((MassiveEquipmentsModificationsInfos) massiveEquipmentsModificationsInfos);
+    }
+
+    private void assignAttributes(MassiveEquipmentsModificationsInfos massiveEquipmentsModificationsInfos) {
+        if (modifications == null) {
+            modifications = massiveEquipmentsModificationsInfos.getModifications().stream().map(ModificationInfos::toEntity).collect(Collectors.toList());
+        } else {
+            modifications.clear();
+            modifications.addAll(massiveEquipmentsModificationsInfos.getModifications().stream().map(ModificationInfos::toEntity).collect(Collectors.toList()));
```
Comment (jonenst) [inline src/main/java/org/gridsuite/modification/server/entities/equipment/modification/MassiveEquipmentsModificationsEntity.java:54]: refactor a bit the common part in this method

File: src/main/java/org/gridsuite/modification/server/dto/MassiveEquipmentsModificationsInfos.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+@SuperBuilder
+@NoArgsConstructor
+@Getter
+@Setter
+@Schema(description = "Massive equipments modifications infos")
+@JsonTypeName("MASSIVE_EQUIPMENTS_MODIFICATIONS")
+@ModificationErrorTypeName("MASSIVE_EQUIPMENTS_MODIFICATIONS_ERROR")
+public class MassiveEquipmentsModificationsInfos extends ModificationInfos {
+    @Schema(description = "equipments modifications")
+    private List<ModificationInfos> modifications;
```
Comment (jonenst) [inline src/main/java/org/gridsuite/modification/server/dto/MassiveEquipmentsModificationsInfos.java:38]: I don't remember, is this enough for our use cases or do we want more structure ? For exemple, grouping multiple modifications in the same "row" of the tabular view ?

```
GeneratorModification(G1, SETX=X1, SETY=Y1)
GeneratorModification(G2, SETX=X2, SETY=Y2)

G1    SETX1  SETY1
G2    SETX2  SETY2
```

But many we want to correlate some other modification with these and we have no hardcoded way of knowing which goes to which ?

```
GeneratorModification(G1, SETX=X1, SETY=Y1)
GeneratorModification(G2, SETX=X2, SETY=Y2)
VoltageLevelModification(VLG1, SETZ=Z1)
VoltageLevelModification(VLG2, SETZ=Z2)

# Do we want this
G1    SETX1  SETY1  SETZ1
G2    SETX2  SETY2  SETZ2

# or is this enough ?
G1      SETX1  SETY1
G2      SETX2  SETY2
VLG1    SETZ1
VLG2    SETZ2
```
Note, if the correlation is always recoverable from the data then we don't need additional structure. For exemple if the data shares an ID, the client can reconstruct the tabular view themselves

File: src/main/java/org/gridsuite/modification/server/ModificationType.java [hunk 4]
```diff
@@ -41,7 +41,8 @@ public enum ModificationType {
     BRANCH_STATUS_MODIFICATION(PreloadingStrategy.NONE),
     DELETE_VOLTAGE_LEVEL_ON_LINE(PreloadingStrategy.NONE),
     DELETE_ATTACHING_LINE(PreloadingStrategy.NONE),
-    GENERATION_DISPATCH(PreloadingStrategy.COLLECTION);
+    GENERATION_DISPATCH(PreloadingStrategy.COLLECTION),
+    MASSIVE_EQUIPMENTS_MODIFICATIONS(PreloadingStrategy.COLLECTION);
```
Comment (geofjamg) [inline src/main/java/org/gridsuite/modification/server/ModificationType.java:45]: Don't call it massive this is not their right term in English. -> maybe TABLE_EQUIPMENT_MODIF

File: src/main/java/org/gridsuite/modification/server/NetworkModificationController.java [hunk 5]
```diff
@@ -196,4 +193,23 @@ public ResponseEntity<Void> deleteLineTypesCatalog() {
         lineTypesCatalogService.deleteLineTypesCatalog();
         return ResponseEntity.ok().build();
     }
+
+    @PostMapping(value = "/network-modifications/voltage-init", consumes = MediaType.APPLICATION_JSON_VALUE)
```
Comment (geofjamg) [inline src/main/java/org/gridsuite/modification/server/NetworkModificationController.java:197]: Voltage init should not be part of this API, the modif is more generic and is just a particular use of generator table motif

**Review by geofjamg — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/modification/server/entities/equipment/modification/MassiveEquipmentsModificationsEntity.java (hunk 2 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/modification/server/entities/equipment/modification/MassiveEquipmentsModificationsEntity.java]: done

File: src/main/java/org/gridsuite/modification/server/NetworkModificationController.java (hunk 5 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/modification/server/NetworkModificationController.java]: renamed

File: src/main/java/org/gridsuite/modification/server/dto/ModificationInfos.java (hunk 1 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/modification/server/dto/ModificationInfos.java:65]: The 'name' properties are no longer necessary here since we added a JsonTypeName annotation on each modification in previous PR

File: src/main/java/org/gridsuite/modification/server/entities/equipment/modification/MassiveEquipmentsModificationsEntity.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+/**
+ * @author Franck Lecuyer <franck.lecuyer at rte-france.com>
+ */
+@NoArgsConstructor
+@Getter
+@Setter
+@Entity
+@Table(name = "MassiveEquipmentsModifications")
+public class MassiveEquipmentsModificationsEntity extends ModificationEntity {
+    @OneToMany(cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/modification/server/entities/equipment/modification/MassiveEquipmentsModificationsEntity.java:35]: This mapping was choosed to take full advantage on existing equipment modification entities, and thus benefit from the existing code when applying the modifications ...

File: src/main/java/org/gridsuite/modification/server/dto/MassiveEquipmentsModificationsInfos.java (hunk 3 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/modification/server/dto/MassiveEquipmentsModificationsInfos.java:38]: I don't think there is a need for grouping multiple modifications in the same row : in that case, the modifications will be shown in different tables

File: src/main/java/org/gridsuite/modification/server/ModificationType.java (hunk 4 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/modification/server/ModificationType.java]: Renamed

File: src/main/java/org/gridsuite/modification/server/NetworkModificationController.java [hunk 7]
```diff
@@ -196,4 +193,23 @@ public ResponseEntity<Void> deleteLineTypesCatalog() {
         lineTypesCatalogService.deleteLineTypesCatalog();
         return ResponseEntity.ok().build();
     }
+
+    @PostMapping(value = "/groups/modification", consumes = MediaType.APPLICATION_JSON_VALUE)
+    @Operation(summary = "Create a group containing a modification")
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "The group with the modification has been created")})
+    public ResponseEntity<UUID> createModificationInGroup(@RequestBody ModificationInfos modificationsInfos) {
```
Comment (sBouzols) [inline src/main/java/org/gridsuite/modification/server/NetworkModificationController.java:200]: Why not adding a entrypoint to create a `groups` resource, then use the existing endpoint to create a `network-modifications` resource ?

File: src/main/java/org/gridsuite/modification/server/entities/equipment/modification/GeneratorModificationEntity.java [hunk 8]
```diff
@@ -215,30 +216,30 @@ public void update(@NonNull ModificationInfos modificationInfos) {
...
-        this.transientReactance = new DoubleModificationEmbedded(generatorModificationInfos.getTransientReactance());
-        this.stepUpTransformerReactance = new DoubleModificationEmbedded(generatorModificationInfos.getStepUpTransformerReactance());
-        this.voltageRegulationType = new EnumModificationEmbedded<>(generatorModificationInfos.getVoltageRegulationType());
-        this.regulatingTerminalId = new StringModificationEmbedded(generatorModificationInfos.getRegulatingTerminalId());
-        this.regulatingTerminalType = new StringModificationEmbedded(generatorModificationInfos.getRegulatingTerminalType());
-        this.regulatingTerminalVlId = new StringModificationEmbedded(generatorModificationInfos.getRegulatingTerminalVlId());
-        this.qPercent = new DoubleModificationEmbedded(generatorModificationInfos.getQPercent());
-        this.reactiveCapabilityCurve = new BooleanModificationEmbedded(generatorModificationInfos.getReactiveCapabilityCurve());
+        this.energySource = generatorModificationInfos.getEnergySource() != null ? new EnumModificationEmbedded<>(generatorModificationInfos.getEnergySource()) : null;
```
Comment (sBouzols) [inline src/main/java/org/gridsuite/modification/server/entities/equipment/modification/GeneratorModificationEntity.java:219]: Why do we need those new null checks ?

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java [hunk 9]
```diff
@@ -251,4 +251,39 @@ public Optional<NetworkModificationResult> duplicateModifications(UUID targetGro
...
     }
+
+    public UUID createModificationInGroup(@NonNull ModificationInfos modificationsInfos) {
+        UUID groupUuid = UUID.randomUUID();
+        networkModificationRepository.saveModifications(groupUuid, List.of(modificationsInfos.toEntity()));
+        return groupUuid;
+    }
+
+    @Transactional
+    public Optional<NetworkModificationResult> duplicateModificationsInGroup(UUID targetGroupUuid,
```
Comment (sBouzols) [inline src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java:262]: We already have this feature when copying nodes full of network-modifications don't we ?

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 10]
```diff
@@ -165,6 +165,9 @@ databaseChangeLog:
   - include:
       file: changesets/changelog_20230702T105416Z.xml
       relativeToChangelogFile: true
+  - include:
```
Comment (sBouzols) [inline src/main/resources/db/changelog/db.changelog-master.yaml:168]: why adding it before changelog_20230728T141452Z, ho I see... hum maybe you should regenerate it ? hum not really useful by the way

**Review by sBouzols — COMMENTED** [review]
some comments will test

File: src/main/java/org/gridsuite/modification/server/NetworkModificationController.java (hunk 7 above)
Comment (sBouzols) [reply src/main/java/org/gridsuite/modification/server/NetworkModificationController.java:200]: I see, there isn't group resource, it's only an UUID, my bad !

File: src/main/java/org/gridsuite/modification/server/entities/equipment/modification/GeneratorModificationEntity.java (hunk 8 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/modification/server/entities/equipment/modification/GeneratorModificationEntity.java:219]: no need to create all those embedded objects with null value and null opType inside ...

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java (hunk 9 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java:262]: I have refactored the code to use a common private method

**Review by sBouzols — APPROVED** [review]
Code Review OK
Tests OK

File: src/main/java/org/gridsuite/modification/server/entities/equipment/modification/VoltageInitModificationEntity.java [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+ * @author Franck Lecuyer <franck.lecuyer at rte-france.com>
+ */
+@NoArgsConstructor
+@Getter
+@Setter
+@Entity
+@Table(name = "voltageInitModification")
+public class VoltageInitModificationEntity extends ModificationEntity {
+    @ElementCollection
+    @CollectionTable(name = "voltageInitGeneratorsModification")
```
Comment (jonenst) [inline src/main/java/org/gridsuite/modification/server/entities/equipment/modification/VoltageInitModificationEntity.java:36]: missing index and foreign key name

note: it's actually missing pretty much everywhere I see...

Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/modification/server/entities/equipment/modification/VoltageInitModificationEntity.java]: Done
